[2단계 - 자동차 경주] 니야(최수연) 미션 제출합니다.

sooyeoniya opened the pull request:
# 🐥 `Step2` 니야의 자동차 경주 리팩토링 🐥
안녕하세요, 도밥! 니야입니다 :)
지난 1단계에서 리뷰해주신 내용을 토대로 2단계 리팩토링 진행해봤습니다 !

도밥이 추천해주신 리뷰 반영표가 가독성이 좋아서 한 번 비슷하게 활용해보고자 아래와 같이 정리해봤습니다 😊

## 📜 리뷰 반영 요약표
| 내용 | 관련 코멘트 | 커밋 |
| -------- | ---------- | -------- | 
| validate는 view일까? domain일까? | [Input 객체 내부에 validate 검증 로직 존재](https://github.com/woowacourse/javascript-racingcar/pull/345#discussion_r1956220120) | https://github.com/woowacourse/javascript-racingcar/pull/402/commits/86d320c0de5ac762d545122bc186bf3a9da60f69 |
| 유지보수성 높이기 위한 필요한 상수화 | [테스트 명세에서 상수 활용하기](https://github.com/woowacourse/javascript-racingcar/pull/345#discussion_r1957033771) | https://github.com/woowacourse/javascript-racingcar/pull/402/commits/e320addc819a3b4fa47958eb24f06a6459d3f1df |
| RaceController 역할의 모호성 | [RaceController가 컨트롤러의 역할은 아닌것 같다.](https://github.com/woowacourse/javascript-racingcar/pull/345#pullrequestreview-2617565741) | https://github.com/woowacourse/javascript-racingcar/pull/402/commits/1bd0768a5126d6194d133a1ab6a555158a69d24c |

리뷰 반영한 것 외에도 자잘한 리팩토링도 좀 많이 진행했습니다. 최대한 커밋 단위 열심히 나누어 놨기 때문에 보시기에는 불편하지 않을 것 같다는 생각이 들긴 하지만,,, 뭔가 궁금하거나 이상한 점 있으면 코멘트 부탁드립니다 ! 😃

## ❓궁금한 점

`RaceController`와 `GameController`를 리팩토링 하면서 풀리지 않는 문제들이 좀 있어서 여쭙고자 합니다.

우선, 요약하자면 이렇습니다.

```
1️⃣ 시도 횟수만큼 반복문을 돌려주는 로직은 도메인 로직인가? 만약 그렇다면 정말로 컨트롤러에 있는 것이 좋은 설계가 아닌 것일까?
2️⃣ `RaceController`에 있던 도메인 로직들을 유틸 함수로 분리했는데, 유틸 함수가 도메인에 전적으로 의존하고 있다면 과연 유틸 함수가 맞을까? 아니면 헬퍼 함수라고 봐야하는 걸까?
3️⃣ 만약, `Race`라는 새로운 도메인을 만들게 되면, `Car`도메인과의 의존성은 어떻게 해결해야 하는 걸까? 
```

사실 어떻게 보면 모든 애플리케이션에서는 도메인과 도메인 사이에 결합이 무조건적으로 존재할 수 밖에 없잖아요. 그렇지 않는다면 모든 기능이 톱니바퀴처럼 잘 맞물려 굴러가지 못할 것이라고 생각합니다. 그 과정에서 도메인과 뷰를 분리하여 서로 상호작용 해주도록 관리해주는 게 컨트롤러인데, 그 분리하는 기준을 잘 모르겠습니다.. 

```js
  #playRace(cars, tryCount) {
    Output.printRaceStart();
    for (let cnt = 1; cnt <= tryCount; cnt++) {
      this.#playRound(cars);
      Console.printLineBreak();
    }
    Output.printWinners(getWinners(cars));
  }
```
위 코드는 `GameController`에 있는 시도 횟수(`tryCount`) 만큼 라운드를 실행해주는 함수입니다. 보면, `Output`이나 `Console` 객체 모듈의 경우에는 모두 `View` 에 해당됩니다. `this.#playRound()`는 내부 함수를 호출하는 것이고, 이는 또 `Domain`과 `View` 의 조합으로 이루어져있습니다. 

### 1️⃣ 시도횟수를 돌리는 위치
여기서 제가 궁금한 것은 `tryCount` 만큼을 반복문 돌려주고 있는 부분입니다. `tryCount`를 컨트롤러에서 직접 조작해도 되는 건지 궁금합니다. (물론 설계 관점에서는 정답은 없다고 생각하지만, 도밥의 의견이 궁금합니다!) `tryCount` 자체가 라운드를 시도횟수만큼 반복해주는 것이기 때문에 도메인 로직이라는 느낌이 들긴 합니다. 그런데 또 막상 따로 들어갈만한 도메인이 존재하지 않습니다. 지금 너무 `MVC` 패턴에 매몰되어 그런걸 수도 있겠지만, 만약 `MVC` 중 하나에 들어가야한다면, 도밥은 이 중 어디에 제일 가깝다고 생각하시나요? 아니면 아예 새로운 계층을 도입해야 하는걸까요?

도밥이 1단계 피드백에서 `니야를 가두는 MVC라는 틀을 과감하게 벗어던지고 view와 도메인 분리에 집중해 보셔요~!` 라고 말씀해주셨는데, 사실 `MVC`라는 것에서 벗어난다고 해도, 이 도메인은 어디에 존재해야할지 모르겠습니다. 제가 `RaceController`를 삭제하면서 점점 미궁에 빠지는 것 같습니다... 😱

프리코스 당시에는 `Race`라는 서비스 계층을 하나 도입하여 자동차 도메인을 의존하는 계층을 따로 두었습니다. 하지만 이번에는 최대한 새로운 계층을 만들지 않고 이 안에서 해결해보고자 유틸 함수로 분리를 했습니다.

### 2️⃣ 유틸 함수
여기서 또 문제인게, 사실  유틸 함수는 특정 작업을 수행하기 위해서 재사용성을 고려한 독립 실행 함수잖아요. 근데 지금 보면 `getWinners`와 `getMaxCount`는 무조건 자동차 인스턴스 배열이 인자로 꼭 들어가야 됩니다. 완전 `Car` 객체에 의존하고 있는 유틸 함수인데, 이게 과연 유틸 함수라고 볼 수 있는 건지 싶습니다.. 그렇다고 `Car` 객체에 넣을 수는 없는 로직이고 (왜냐하면 `Car` 객체는 하나의 자동차에 대한 인스턴스를 생성하는데, 이 유틸들은 모든 자동차들을 한 번에 계산해야 하기 때문!), 컨트롤러에도 존재할 수 없는 도메인 로직이기 때문입니다. (이것 때문에 `RaceController` 역할의 모호성에 대한 피드백을 받은 것이구요.) 그래도 물론 `RaceController` 때보다는 훨씬 단위 테스트가 수월해졌기 때문에 전보다는 훨씬 깔끔하다는 생각이 들긴 합니다. (`RaceController` 테스트는 사실 통합 테스트에 가깝기도 했고, 그 내부에서 모킹 함수를 사용하고 있었기 때문입니다.)

### 3️⃣ Race 도메인의 위치
만약, `Race`라는 새로운 도메인을 만들게 되면, 도메인 간의 의존성은 어떻게 해결되는 건가요? `Race`자체는 `Car`라는 객체에 전적으로 의존할 수 밖에 없다고 생각합니다. 그러면 `Race`는 어느 폴더에 위치하는 것이 가장 이상적인 걸까요? 이 과정에서 도메인과 도메인 간의 결합이나 의존을 최대한 끊어내고 응집도를 높이는 방법은 따로 없을까요? 
더 좋은 디자인 패턴이나 구조가 있다면 추천해주시면 정말 감사할 것 같습니다!

제가 너무 정답 없는 추상적인 질문만 드리는 것 같아서 죄송하기도 하고, 그래서 더 문장 정리가 제대로 안되는 것 같아서 좀 아쉽지만, 함께 의견 나눠보고 싶어서 작성해봤습니다! 제가 잘 몰랐거나 놓친 부분이 있다면 말씀해주시면 감사하겠습니다 😀

sooyeoniya committed:
refactor: models 폴더 domains로 변경

sooyeoniya committed:
refactor: 전체 메소드명, 변수명 적절하게 수정

- 단복수형 맞추고, 메서드의 경우 동사로 시작하도록 수정

sooyeoniya committed:
refactor: validate(domain)와 Input(UI) 분리 및 재사용 코드 분리

- Input 값이 유효하지 않은 경우 재입력 받는 로직 유틸 함수로 분리
- 재귀 호출 방식이 아닌 while 문으로 변경 -> 무한재귀로 인한 스택오버플로우 방지

sooyeoniya committed:
refactor(validation): 자동차 이름 입력시 구분자(,)로 인해 빈 값이 나오는 경우 무시

- 자동차 이름 배열의 길이가 1이면서 빈 값이거나, 구분자(,)로 인해 배열에 빈 값이 들어있는 경우, 빈 값을 filter로 없애고 유효성 검증 하도록 수정함.
- 어차피 자동차가 2대 미만인 경우에 에러가 발생하기 때문에 배열이 완전 빈 값인 경우 따로 예외 처리를 안해줘도 괜찮고,
- 사용자의 구분자(,) 입력 실수는 UX 측면에서 봤을 때 에러 발생보다는 그냥 정상적으로 처리해주는 것이 맞다고 판단함.

sooyeoniya committed:
refactor: printWinners 메서드 단수->복수형으로 변경

sooyeoniya committed:
refactor(validation): 유효성 검증 함수 리팩토링

- 유지보수 및 가독성 측면에서 더 직관적인 코드가 되도록 함수 분리함.
- 매직 넘버로 시도횟수 최솟값, 최댓값 명명.

sooyeoniya committed:
refactor(RaceController): getWinners 메서드에서 최댓값 추출하는 로직 분리

- 함수가 하나의 일을 하도록 분리

sooyeoniya committed:
chore: 시도 횟수 테스트 케이스 명세 통일

- 미만, 초과 / 이상, 이하 로 묶어서 통일시키기

sooyeoniya committed:
test: 행위 주도 개발 방식으로 테스트 명세 수정

- 우리가 개발하고자 하는 것은 어떻게(how)보다 무엇(what)이 더 중요함.
- count라는 변수명이 바뀔 수도 있으므로 이런 방식보다는 자동차의 움직임, 즉 행위를 위주로 서술하는 것이 좋음.

sooyeoniya committed:
docs: 행위 주도 개발(BDD)에 대한 코드 문제 및 해결 내용 정리

sooyeoniya committed:
refactor(constants): 중복되는 매직 넘버 상수화

- 의미가 있는 숫자들 상수로 변경
- 여러 곳에서 사용되는 경우 constants에서 선언
- EXPRESSION_FLAG 의 경우에는 Output에서만 사용되므로 응집도 향상을 위해 Output의 printRace 함수 내부로 이동
- 전반적으로 유지보수성 높이는 데 기여함.

sooyeoniya committed:
refactor: getRandomNumber 재사용 가능하도록 범위 설정 값 매개변수 추가

sooyeoniya committed:
docs: 재귀 호출과 while 문의 차이점 정리

sooyeoniya committed:
refactor: RaceController 내부 로직들 역할에 맞게 리팩토링

sooyeoniya committed:
chore: 전진 거리를 distance -> count 로 통일

- distance 로 하려니까 통일성이 없어짐. 그냥 다시 count로 통일시키기. car 객체 내부에서 count라고 명명하고 있으므로.

sooyeoniya committed:
test: getMaxCount 함수 테스트 구현

sooyeoniya committed:
test: getWinners 함수 테스트 구현

sooyeoniya committed:
docs: RaceController 문제 분석 및 리팩토링 과정 문서 정리

sooyeoniya committed:
docs: README.md Step2 리팩토링 리스트 정리

sooyeoniya committed:
feat(utils): 조건문에 따른 에러 반환 로직 유틸로 분리

- 유효성 검증하는 모듈에서 여러번 재사용되고 있어 분리함.

Creative-Lee commented on src/controllers/GameController.js:
```diff
@@ -1,24 +1,41 @@
+    cars.forEach((car) => {
+      car.move(getRandomNumber(0, 9));
+      Output.printRace(car.name, car.count);
+    });
```
1. domain과 domain간의 의존은 잠시 뒷일로 미뤄도 좋으니, 니야가 생성을 고려했던 Race.js를 만들어 봅시다.
2. 그리고 니야가 도메인 로직이라고 생각했던 위 로직들을 Race.js로 분리해봅시다. 
3. 과정에서 분리할 수 없다고 생각했던 view 로직의 분리는 본문에 남겨드린 코멘트를 참고해서 분리해봅시다.

Creative-Lee commented on src/utils/getMaxCount.js:
```diff
@@ -0,0 +1,13 @@
+import Car from "../domains/Car.js";
+
+/**
+ * 자동차들의 전진 거리 중 최댓값을 반환한다.
+ * @param {Array<Car>} cars
+ * @returns {number}
+ */
+const getMaxCount = (cars) => {
+  const countArray = cars.map((car) => car.count);
+  return Math.max(...countArray);
+};
+
+export default getMaxCount;
```
이미 니야가 알고있던 것 처럼, util로 분리한 승자를 구하는 로직들을 다시 도메인 클래스가 가지도록 분리해봅시다.

Creative-Lee reviewed (changes requested):
안녕하세요 니야~! step2에서 꽤나 많은 변화가 있었네요 ㅎㅎ
궁금한 점 적어주신 것 모두 **니야의 생각이 잘 담겨있어서 너무 좋네요.**
'이렇게 생각하기 때문에 이렇다고 생각합니다.' 라는 식의 질문이라 답변도 수월했습니다😄 
자 그럼 천천히 하나씩 궁금증을 해결해봅시다. 

---
 
> ###  1️⃣ 시도횟수를 돌리는 위치

> tryCount를 컨트롤러에서 직접 조작해도 되는 건지 궁금합니다. tryCount 자체가 라운드를 시도횟수만큼 반복해주는 것이기 때문에 도메인 로직이라는 느낌이 들긴 합니다. 

결론부터 말씀드리면 저 역시 domain 로직이라고 생각해요. 가능하다면 domain 클래스에서 호출하면 좋겠죠. 

> 그런데 또 막상 따로 들어갈만한 도메인이 존재하지 않습니다.

제가 step1에서 니야에게 'RaceContoller가 컨트롤러의 역할은 아닌것 같다.' 라고 드렸던 코멘트의 의미는 RaceController 내부에 경주를 하고, 각 라운드를 진행하고, 승자를 계산하는 로직이 있었기 때문이에요. 저는 이것들이 모두 도메인 로직으로 생각되었어요.
때문에 controller라는 디렉터리에 RaceContoller 라는 이름을 가지고 있는게 어색하게 느껴졌구요. 
다시 말해 디렉터리와 이름만 바뀌었어도 그 자체로 이미 훌륭한 domain 클래스라고 부를수 있었다는 말입니다.(view 로직과의 결합이 존재하는것은 별개의 일이고요)

>컨트롤러는 도메인과 뷰의 상호작용 역할을 해주어야 한다고 생각했다.
그렇기 때문에 `GameController`에서는 로직 상 떼어낼 수 없는 뷰 부분을 관리하고, 최대한 다른 도메인 모듈을 가져다 쓸 수 있도록 리팩토링 했다.

니야의 트러블 슈팅 문서 일부인데요..! 조심스럽게 추측해보자면, 
`GameController`에 일부로직에 도메인 로직과 떼어낼 수 없는 뷰 가 존재했고, 이 모습이 니야가 controller의 역할이라고 정의한 개념과 비슷하다고 생각했을 것 같아요. 

**근데 '만약 떼어낼 수 없는 뷰' 라고 생각했던 부분이 떼어내 졌다면 어떤가요?**
니야가 떼낼수 없다고 생각했던 view로직은 아마도 라운드 결과를 출력하는 코드였던것 같습니다.

```ts
  #playRace(cars, tryCount) {
    Output.printRaceStart();
    for (let cnt = 1; cnt <= tryCount; cnt++) {
      this.#playRound(cars);
      Console.printLineBreak(); // 이 부분
    }
    Output.printWinners(getWinners(cars));
  }
```
라운드 결과를 배열에 모아두었다가, 외부에서 한번에 출력하는 방법
외부에서 라운드가 끝날때마다 실행할 함수(미션에서는 출력함수겠죠)를 인자로 넘겨받는 방법 등이 있었을것 같아요. 

저는 니야가 이 떼어낼 수 없는 뷰를 떼어내고 최대한 도메인 로직을 분리하는것에 포커스를 맞췄다면 어땠을까? 생각했어요. 니야가 충분히 해낼 수 있었을 거라고도 생각하고요.

> 프리코스 당시에는 Race라는 서비스 계층을 하나 도입하여 자동차 도메인을 의존하는 계층을 따로 두었습니다. 하지만 이번에는 최대한 새로운 계층을 만들지 않고 이 안에서 해결해보고자 유틸 함수로 분리를 했습니다.

도메인을 의존하는 도메인을 만들고 싶지 않아서 새로운 시도를 한것으로 이해했어요. 더 나은 구조를 만들기 위해 좋은 시도를 했다고 생각해요. 

--- 

> ### 2️⃣ 유틸 함수

다만 도메인 로직을 유틸로 분리했다고 해서 도메인이 아닌것은 아니겠죠. 
이미 니야의 질문속에 답이 있어요. 

---

> ### 3️⃣ Race 도메인의 위치

> Race자체는 Car라는 객체에 전적으로 의존할 수 밖에 없다고 생각합니다. 그러면 Race는 어느 폴더에 위치하는 것이 가장 이상적인 걸까요? 이 과정에서 도메인과 도메인 간의 결합이나 의존을 최대한 끊어내고 응집도를 높이는 방법은 따로 없을까요?
더 좋은 디자인 패턴이나 구조가 있다면 추천해주시면 정말 감사할 것 같습니다!

저는 이 미션을 할때 도메인과 도메인의 의존을 끊으려는 생각조차 못했던것 같아요😅 
그저 domain과 view를 분리하는것에 집중했던것 같아요. 
(그런 부분에서 니야가 대단하다고 생각해요!)

또, 디자인 패턴을 따로 공부해본적이 없습니다. 그래서 패턴이나 구조는 추천할 수 없을것 같아요.(너무 어렵다는 선입견을 가지고 있기도 하고, 실제로 어려운것 같고요.)

그래도 생각나는 대로 답변드려보자면,
**Race가 관리하는 객체를 딱! 정해서 Car! 라고 정의하지 않는 것**부터 시작해볼것 같아요.
우선 RaceGame이 말(player)로 사용할 수 있는 클래스를 받아 생성하도록 개선해볼것 같아요.
이 말(player)은 필드로 position과 name을 가지고, move라는 메서드를 가지면 되겠네요. 
이 구성을 만족했다면 그것이 자동차이던, 기차이던, 사람이던 뭐든 다 RaceGame을 구동하는데 사용될 수 있겠네요. 최소한 RaceGame이 직접 `말.js`를 import 하지 않을 수 있겠네요.

---

음.. 조금 조심스럽지만, 솔직하게 말씀드려 볼게요.
질문에 답변하면서 니야가 코드에 대해 자신이 없어 하는것을 느꼈어요.
질문에 답이 수월했던 결정적인 이유는 **이미 대부분의 내용을 잘 알고계시기 때문이에요.**
다만, 어딘가 정답을 찾는듯한 모습이 보여서 조금은 걱정이 되네요.
니아! 코드에 대해 자신감을 더 가져도 좋을 것 같아요!!
코멘트에 제 생각을 남겨 놓았으니 확인해보시고 재요청주세요!

sooyeoniya committed:
refactor: Race 객체 생성 및 View 분리

- 각 라운드 출력 로직을 Race 도메인과 분리하여 Controller로 넘겨 View에게 실행하도록 수정함
- 사용하지 않는 유틸 및 테스트 코드 삭제

sooyeoniya committed:
chore: 랜덤 숫자 범위 최솟값, 최댓값 상수화

sooyeoniya committed:
chore: 전체 경주 결과 출력 함수명 변경

- printRoundResults -> printRaceResults로 변경

sooyeoniya committed:
test: Race 클래스 테스트 구현

- 모킹하지 않는 선에서 구현

sooyeoniya committed:
docs: README.md Step2 리팩토링 v2 리스트 및 내용 작성

- README.md 에 한 일 목록 작성
- refactor-step2-v2.md 에 리팩토링에 대한 상세 내용 작성

sooyeoniya commented:
# 🐥 `Step2` 니야의 자동차 경주 리팩토링 - 도밥 피드백 반영 v2 🐥

도밥! 리뷰 남겨주신 것 꼼꼼히 읽어봤습니다. 확실히 도밥이 추천해준대로 다시 생각해보니까 굳이 분리 안할 이유가 없더라구요!!
그래서 이번에는 도메인 의존성에 대해서 너무 깊게 고민하지 않고, 우선 Race 객체를 하나 만들어서 View와 명확히 분리하는 방식으로 리팩토링 해보았습니다! 훨씬 더 깔끔한 코드가 된 것 같아서 기분 좋아졌습니다 :)

## ✅ `Race` 객체 생성 및 `View` 분리

- 도메인과 뷰를 명확히 분리하는 작업을 통해 `Controller`에서는 `Output` 모듈만 사용되고 `Console` 사용을 없앨 수 있었습니다! 
- 즉, `Output`과 `Input` 내부에서만 `Console`을 사용하도록 리팩토링되어서, 코드가 훨씬 깔끔해졌다고 느꼈습니다.

## ✅ `Race` 클래스 테스트에서 모킹 함수 사용하지 않는 방법
어쩔 수 없이 `Race` 클래스의 내부에 `getRandomNumber` 함수가 존재할 수 밖에 없어서, 이 함수의 모킹을 어떻게 하면 피할 수 있을지 고민해봤습니다. 일단 `public` 함수는 아래 2개가 있고, 각각 테스트를 어떻게 진행했는지 작성하였습니다!

### ✔️ `Race` 클래스의 `playRace` 함수

`playRace`를 테스트 하려면 어쩔 수 없이 모킹 함수를 사용해야 했는데요. (`playRace`에서는 `playRound`를 호출하고, `playRound`에서는 `getRandomNumber` 함수를 호출하고 있기 때문)
만약 모킹을 하지 않고 테스트하려면, `랜덤으로 인한 결과 값이 고정되지 않는다는 전제` 하에, 다른 매처들을 활용해 간접적으로 테스트를 진행하는 방식을 채택해야 했습니다.

  - 예를 들면, `winners` 배열이 1명 이상 꼭 배출되는 것을 확인 하기 위해 `toBeGreaterThan`를 사용한다든지.
  - 매개변수로 입력 받은 `tryCount`에 따라 라운드 개수, 즉 `playRound`가 `tryCount`만큼 호출된다든지. 사실 이 방법은 `playRound`가 `public` 함수거나 모킹 함수일때만 가능하므로, `playRace` 반환 배열의 길이를 확인하면 된다고 생각하여 그렇게 테스트를 진행했습니다.

### ✔️ `Race` 클래스의 `getWinners` 함수

`getWinners` 함수는 랜덤 숫자를 생성하는 함수만 피해서 테스트하면 된다고 생각했습니다.
따라서, 각 자동차 인스턴스의 `move`를 직접 하나씩 호출하여 해당 인스턴스를 사용한 `Race` 객체 인스턴스를 생성한 후,(이렇게 되면 최종 우승자를 예상할 수 있기 때문입니다.) `playRace`를 돌리지 않고 바로 `getWinners`를 통해 우승자를 계산해 반환하는 로직만 테스트 하도록 구현해봤습니다.

---
`PS.` 그리고 도밥의 마지막 한마디를 보고 생각을 좀 해봤는데요. 사실 도밥 말처럼 제 코드에 대한 자신감이 없었던 걸지도 모른다는 생각이 들었습니다. 하지만, 뭔가 자신감 문제라기보다는 어떻게 하면 제 맘에 드는 더 좋은 깔끔한 설계가 될 수 있을지, 그리고 리뷰어에게 도움을 받아 더 좋은 방식을 추천받고 싶어서 계속 고민하고 PR을 작성했던 것 같아요. 근데 코드라는 것 자체가 좋은 설계는 있을 수 있지만, 일단 개성이 담겨있기도 하고 정답은 없으니까 도밥 말씀대로 `정답`만 쫓는 건 지양하도록 해야겠어요! 앞으로는 저 자신을 더 믿어보고 다양한 시도를 해보며 발전할 수 있는 `니야`가 되도록 노력하겠습니다 ! 좋은 피드백 감사해요 😊

Creative-Lee commented on src/domains/Race.js:
```diff
@@ -0,0 +1,42 @@
+    this.#cars = cars;
+  }
+
+  /**
+   * 전체 경주 실행 및 전체 라운드 결과 반환
+   * @param {number} tryCount
+   * @returns {Array<Array<{ name: string, count: number }>>}
+   */
```
타이핑까지 완벽하네요 👍

Creative-Lee commented on src/domains/Race.js:
```diff
@@ -0,0 +1,42 @@
+   * @param {number} tryCount
+   * @returns {Array<Array<{ name: string, count: number }>>}
+   */
+  playRace(tryCount) {
+    return Array.from({ length: tryCount }, () => this.#playRound());
+  }
```
배열 메서드 활용도 좋습니다!

Creative-Lee reviewed (approved):
>도밥! 리뷰 남겨주신 것 꼼꼼히 읽어봤습니다. 확실히 도밥이 추천해준대로 다시 생각해보니까 굳이 분리 안할 이유가 없더라구요!!
그래서 이번에는 도메인 의존성에 대해서 너무 깊게 고민하지 않고, 우선 Race 객체를 하나 만들어서 View와 명확히 분리하는 방식으로 리팩토링 해보았습니다! 훨씬 더 깔끔한 코드가 된 것 같아서 기분 좋아졌습니다 :)

기분이 좋아지셨다니 아주 좋네요 👍 
니야가 탐구했던 도메인과 도메인의 결합에 대한 부분도 Car를 Race에서 바로 import하는게 아니라 초기화 시 인자로 받도록 구현하셔서 결합이 비교적 덜하게 되었네요 💯
사실상 미션을 마무리해도 충분히 좋은 코드이지만, 아쉬우실수도 있으니 니야의 고민에 대한 의견을 드려볼게요.

---

> playRace 반환 배열의 길이를 확인하면 된다고 생각하여 그렇게 테스트를 진행했습니다.
> getWinners 함수는 랜덤 숫자를 생성하는 함수만 피해서 테스트하면 된다고 생각했습니다.

어떻게든 검증할 수 부분이라면 최대한 검증해주셨다고 생각해요. 
간접적으로 테스트 하는 방향으로 구상해서 잘 작성해주셨어요. 할 수 있는 최선을 다하신게 코드와 글에서 충분히 느껴집니다! 
사실 이대로도 이 미션에서는 충분하지만, 

>✔️ Race 클래스의 playRace 함수
>playRace를 테스트 하려면 어쩔 수 없이 모킹 함수를 사용해야 했는데요. (playRace에서는 >playRound를 호출하고, playRound에서는 getRandomNumber 함수를 호출하고 있기 때문)

이 어쩔수 없었던 부분이 니야에게는 찝찝한 숙제로 남아있을 것 같은데요~!
지금의 구조에서 random 함수가 포함된 Race를 어떻게 테스트 해볼 수 있을까요? 
숫자를 생성하는 함수를 외부에서 인자로 받는 방법을 사용해 볼 수 있을것 같아요.
Random 함수는 default parameter로 두어서 사용처에서는 아무런 수정없이 그대로 로직을 실행하도록 하고요.
테스트 작성시에는 내가 원하는 숫자를 뱉는 작은 유틸함수를 넣어 원하는대로 조작해서 테스트 할 수 있을것 같아요.
jest.fn 같은 mocking을 활용하는것과 비슷하게 느껴질 수 있지만, 비교적 독립적이고 예측 가능한 테스트를 작성할 수 있겠죠.
저는 자동차 경주 미션 ~ 로또 미션까지 LMS와 우테코 블로그, 그리고 [이 글](https://jojoldu.tistory.com/676?category=1036934)에서 도움을 많이 받았습니다.

코드 수정 요청 드릴 부분은 없어서 Approve처리를 해두고 마지막 자유?시간을 드리려고해요.
코멘트 읽어보시고 (반영을 원하시면 추가로 반영하시고) DM이나 멘션으로 알림만 주셔요~~

우테코 첫미션 고생 정말 많으셨습니다 니야~!
덕분에 저도 많이 얻어갑니다👍 
다음에 또 봐요~ 👋
